File: docs-web.resources/com.ibm.docs.web.resources/WebContent/js/concord/chart/model/Axis.js
/* ***************************************************************** */
/*                                                                   */
/* IBM Confidential                                                  */
/*                                                                   */
/* IBM Docs Source Materials                                         */
/*                                                                   */
/* (c) Copyright IBM Corporation 2012. All Rights Reserved.          */
/*                                                                   */
/* ***************************************************************** */

dojo.provide("concord.chart.model.Axis");
dojo.require("concord.chart.model.Title");
dojo.require("concord.chart.utils.Utils");

dojo.declare("concord.chart.model.Axis", null,{	
	id: null,
	type: null,
	position: null, //left, right, top, bottom
	invisible : null, 
	tickLblPos: null,   //label position, if none, then the label will not show
	textRotation: 0,
	
	//scaling
	logBase : -1,      //10 ....
	max : null,
	min : null,
	majorUnit : null,
	minorUnit : null,
	
	//grid properties
	majorGrid : null,     //GraphicPro
	minorGrid : null,     //GraphicPro
	
	//number format
	format : null,
	sourceLinked : 1,
	preserveCode : null,
	
	//axis title
	title : null,
	
	spPr : null,
	textPro : null,
	
	categorySeq : null,   //DataSequence
	
	changes: 0,
	utils : concord.chart.utils.Utils,
	
	constructor: function(chart)
	{
		this._chart = chart;
	},
	
	toJson: function(noData)
	{
		var json = {};
		json["axId"] = this.id;
		if(this.type)
			json["type"] = this.type;
		json["delete"] = this.invisible ? 1:0;
		json["axPos"] = this.position;
		
		if(this.tickLblPos!=null)
			json["tickLblPos"] = this.tickLblPos;
		
		var scaling = {};
		if(this.logBase>0)
			scaling["logBase"] = this.logBase;
		if(this.max!=null)
			scaling["max"] = this.max;
		if(this.min!=null)
			scaling["min"] = this.min;
		json["scaling"] = scaling;
		
		if(this.majorUnit>0)
			json["majorUnit"] = this.majorUnit;
		if(this.minorUnit>0)
			json["minorUnit"] = this.minorUnit;
		
		if(this.textRotation != 0)
			json["txRot"] = this.textRotation;
			
		if(this.majorGrid!=null)
			json["majorGridlines"] = this.majorGrid;
		if(this.minorGrid!=null)
			json["minorGridlines"] = this.minorGrid;
		if(this.format!=null)
		{
			var numFmt = {};
			numFmt["format"] = this.format;
			numFmt["sourceLinked"] = this.sourceLinked;
			if(this.preserveCode)
				numFmt["pCode"] = this.preserveCode;
			json["numFmt"] = numFmt;
		}
		
		if(this.title!=null)
			json["title"]  = this.title.toJson();
		if(this.spPr!=null)
			json["spPr"] = dojo.clone(this.spPr);
		if(this.textPro!=null)
			json["txPr"] = dojo.clone(this.textPro);
		
		if(this.categorySeq!=null)
		{
			var category = {};
			var ref = this.categorySeq.getAddress();
			if(ref && ref.length > 0)
				category.ref = this.categorySeq.getAddress();
			var data =  this.categorySeq.getData();
			if(data != null){
				category.cache = {};
				category.cache.ptCount = data.length;
				if(!noData || !category.ref)
					category.cache.pts = data;
			}
			var exRef = this.categorySeq.getExRef();
			if(exRef && exRef.length > 0)
				category.exRef = exRef;
			json["cat"] = category;
		}
		return json;
	},
	
	loadFromJson: function(json)
	{
		this.id = json["axId"];
		this.type = json["type"];
		this.invisible = json["delete"];
		this.position = json["axPos"];
		this.tickLblPos = json["tickLblPos"];
		
		var scaling = json["scaling"];
		if(scaling!=null)
		{
			if(scaling["logBase"]!=null)
				this.logBase = scaling["logBase"];
			if(scaling.max!=null)
				this.max = scaling.max;
			if(scaling.min!=null)
				this.min = scaling.min;
		}
		
		if(json["majorUnit"]!=null)
			this.majorUnit = json["majorUnit"];
		if(json["minorUnit"]!=null)
			this.minorUnit = json["minorUnit"];
		
		var rot = json["txRot"];
		if(rot!=null)
			this.textRotation = rot;
		
		this.majorGrid = json["majorGridlines"];
		this.minorGrid = json["minorGridlines"];
		
		var numFmt = json["numFmt"];
		if(numFmt!=null)
		{
			this.format = numFmt["format"];
			if(numFmt["sourceLinked"]!=null)
				this.sourceLinked = numFmt["sourceLinked"];
			this.preserveCode = numFmt["pCode"];
		}
		
		var titleJson = json["title"];
		if(titleJson!=null)
		{
			this.title = new concord.chart.model.Title();
			this.title.loadFromJson(titleJson);
		}
		
		this.spPr = json["spPr"];
		this.textPro = json["txPr"];
		
		var categoriesJson = json["cat"];
		if(categoriesJson!=null)
		{
			var pts = null;
			if(categoriesJson.cache)
				pts = categoriesJson.cache.pts;
			var exRef = categoriesJson.exRef;
			this.categorySeq = this._chart._dataProvider.createDataSequence({ref:categoriesJson.ref, "role" : "cat", "pts" : pts, "exRef" : exRef});
			this.categorySeq.setProperty("role", "cat");
			this.categorySeq.setParent(this);
		}
	},
	
	/*
	 * def: formatted by dojo;
	 * value: original value;
	 */
	formatting: function(def, value, precision)
	{
		dojo.require("concord.chart.i18n.Number");
		
		if (typeof value == "number" && (!this.format || this.format.cat == "text")) {
			def = concord.chart.i18n.Number.roundFloat(value);
		}

		if(!this.format || this.format.cat == "boolean" || this.format.cat == "text")
			return def;

		if(concord.chart.i18n.Number.isDefaultFormat(this.format))
			return def;

		return concord.chart.i18n.Number.format(value, this.format);
	},
	
	set: function(json)
	{
		if("cat" in json)
		{
			if(this.categorySeq!=null)
				this.categorySeq.destroy();
			this.categorySeq = null;
			this.changes |= 0x01;
			
			this._chart.controller.checkData = true;
			var catJson = json["cat"];
			if(catJson)
			{
				var pts = null;
				if(catJson.cache)
					pts = catJson.cache.pts;
				
				var dataProvider = this._chart._dataProvider;
				this.categorySeq = dataProvider.createDataSequence({ref: catJson.ref, "role" : "cat", "pts" : pts});
			
				this.categorySeq.setProperty("role", "cat");
				this.categorySeq.setParent(this);
			}
		}
		
		if("spPr" in json)
		{
			if(this.spPr==null)
				this.spPr = json["spPr"];
			else
			{
				if(json["spPr"]==null)
					this.spPr = null;
				else
					this.utils.mergeSpPr(this.spPr,json["spPr"]);
			}
		}
		
		if("scaling" in json)
		{
			var scaling = json["scaling"];
			if(!scaling)
			{
				this.max = null;
				this.min = null;
			}
			else
			{
				if("min" in scaling)
					this.min = scaling["min"];
				if("max" in scaling)
					this.max = scaling["max"];
			}
		}
		if("majorUnit" in json)
			this.majorUnit = json["majorUnit"];
		
		if("minorUnit" in json)
			this.minorUnit = json["minorUnit"];
		
		if("title" in json)
		{
			var titlePro = json["title"];
			if(titlePro==null)
				this.title = null;
			else
			{
				if(this.title==null)
				{
					this.title = new concord.chart.model.Title();
					this.title.loadFromJson(titlePro);
				}
				else
				{	
					this.title.set(titlePro);
				}
			}
		}
		
		if("txPr" in json)
		{
			if(this.textPro==null)
				this.textPro = json["txPr"];
			else
			{
				if(json["txPr"]==null)
					this.textPro = null;
				else
					dojo.mixin(this.textPro,json["txPr"]);
			}
		}
	},
	
	getReverseSettings: function(pro)
	{
		var undoPro = {};
		if(!pro)
			return undoPro;
		
		if("cat" in pro)
		{
			if(this.categorySeq==null)
				undoPro.cat = "";
			else
			{
				var address = this.categorySeq.getAddress();
				if(address)
					undoPro.cat = {ref:address};
				else
				{
					var data = this.categorySeq.getData();
					undoPro.cat = {cache:{"pts":data}};
				}
			}
		}
		
		if("spPr" in pro)
		{
			if(this.spPr==null)
				undoPro.spPr = null;
			else
				undoPro.spPr = this.utils.reverseSpPr(pro.spPr,this.spPr);
		}
		
		if("txPr" in pro)
		{
			if(this.textPro==null)
				undoPro.txPr = null;
			else
				undoPro.txPr = this.utils.reverseTxPr(pro.txPr,this.textPro);
		}
		
		if("scaling" in pro)
		{
			var scaling = pro["scaling"];
			if(scaling)
			{
				undoPro.scaling = {};
				if("min" in scaling)
					undoPro.scaling.min = this.min;
				if("max" in scaling)
					undoPro.scaling.max = this.max;
			}
		}
		
		if("majorUnit" in pro)
			undoPro.majorUnit = this.majorUnit;
		
		if("minorUnit" in pro)
			undoPro.minorUnit = this.minorUnit;
		
		if("title" in pro)
		{
			if(this.title==null)
				undoPro.title = null;
			else
			{
				undoPro.title = this.title.getReverseSettings(pro.title);
			}
		}
		
		return undoPro;
	},
	
	isCateDirty: function()
	{
		return (this.changes & 0x01) || (this.categorySeq && this.categorySeq.isDirty());
	},
	
	setCategorySeq: function(seq)
	{
		if(this.categorySeq)
			this.categorySeq.destroy();
		
		this.categorySeq = seq;
		if(seq!=null)
			this.categorySeq.setParent(this);
	}
});

File: conversion/com.ibm.symphony.conversion.service.common/src/com/ibm/symphony/conversion/service/common/chart/Axis.java
package com.ibm.symphony.conversion.service.common.chart;

import com.ibm.json.java.JSONObject;

public class Axis
{
	public String id;
	public String position;  //left, right, top, bottom
	public float  crossesAt = 0; //axis cross value
	public String dimension; //x, y
	public int delete = 0;   //0: visible, 1: invisible
	public boolean displayLabel = true;
	
	//scaling
	public String orentation;  //minmax, maxmin
	public float logBase = -1;      //10 ....
	public Double max;
	public Double min;
	public float majorUnit = -1;
	public float minorUnit = -1;
	
	public Float rotationAngle;
	
	//grid properties
	public Style.Graphic majorGrid;
	public Style.Graphic minorGrid;
	
	//number format
	public ChartNumberFormat format;
	public Boolean sourceLinked = true;
	
	//axis title
	public Title title;
	
	public Style.Graphic graphicPro;
	public Style.Text textPro;
	
	public DataReference categoriesRef;
	
	public PlotArea parent;
	
	public int changes = 0;
	
	public ChartDocument chart;
	
	public Axis(ChartDocument chartDoc)
	{
		chart = chartDoc;
	}
	
	public JSONObject toJson()
	{
		JSONObject json = new JSONObject();
		//Some special odf file has no id
		if(id==null || id.length()==0)
			id = "x".equals(dimension) ? "primary-x" : "primary-y";
		json.put("axId", id);
		json.put("delete", delete );
		
		if(!displayLabel)
		  json.put("tickLblPos", "none" );
		
		Boolean vertical = parent.getVertical();
		if(vertical!=null && vertical)
		{
			if("x".equals(dimension))
				dimension = "y";
			else
				dimension = "x";
		}
		if("x".equals(dimension))
		{
			if("end".equals(position))
				json.put("axPos", "t");
			else if("start".equals(position))
				json.put("axPos", "b");
			else
			{
			  if("primary-x".equals(id) || "primary-y".equals(id))
			    json.put("axPos", "b");
			  else
			    json.put("axPos", "t");
			  if(position!=null)
			    crossesAt = Float.parseFloat(position);
			}
		}
		else
		{
			if("end".equals(position))
				json.put("axPos", "r");
			else if("start".equals(position))
				json.put("axPos", "l");
			else
			{
			  if("primary-y".equals(id) || "primary-x".equals(id))
                json.put("axPos", "l");
              else
                json.put("axPos", "r");
              if(position!=null)
                crossesAt = Float.parseFloat(position);
			}
		}
		json.put("crossesAt", crossesAt);
		if(rotationAngle!=null)
		  json.put("txRot", rotationAngle);
		
		JSONObject scaling = new JSONObject();
		if(orentation!=null)
			scaling.put("orientation", orentation);
		if(logBase>0)
			scaling.put("logBase", logBase);
		if(max!=null)
			scaling.put("max", max);
		if(min!=null)
			scaling.put("min", min);
		if(!scaling.isEmpty())
			json.put("scaling", scaling);
		
		if(majorUnit>0)
			json.put("majorUnit", majorUnit);
		if(minorUnit>0)
			json.put("minorUnit", minorUnit);
		
		if(majorGrid!=null)
			json.put("majorGridlines", majorGrid.toJson());
		if(minorGrid!=null)
			json.put("minorGridlines", minorGrid.toJson());
		
		JSONObject numFmt = new JSONObject();
		if(format != null)
		  numFmt.put("format", format.toJson());
		if(sourceLinked)
		  numFmt.put("sourceLinked", 1);  //use data source format
		else
		  numFmt.put("sourceLinked", 0);  //use axis format
		
		json.put("numFmt", numFmt);
		
		
		if(title!=null)
			json.put("title", title.toJson());
		if(graphicPro!=null)
			json.put("spPr", graphicPro.toJson());
		if(textPro!=null)
			json.put("txPr", textPro.toJson());
		
		if(categoriesRef!=null)
			json.put("cat", categoriesRef.toJson());
		
		return json;
	}
	
	public void loadFromJson(JSONObject content)
	{
	  id = (String)content.get("axId");
      delete = ((Number)content.get("delete")).intValue();

      String tickLblPos = (String)content.get("tickLblPos");
      displayLabel = !"none".equals(tickLblPos);
      
      String axPos = (String)content.get("axPos");
      Boolean vertical = parent.getVertical();
      boolean beVertical = vertical!=null && vertical;
	  if("t".equals(axPos))
	  {
		  dimension = beVertical ? "y":"x";
		  position = "end";
		  this.parent.putAxisIdPair(id, beVertical ? "secondary-y":"secondary-x");
	  }
	  else if("b".equals(axPos))
	  {
		  dimension = beVertical ? "y":"x";
		  position = "0";
		  this.parent.putAxisIdPair(id, beVertical ? "primary-y":"primary-x");    		  
	  }
	  else if("l".equals(axPos))
	  {
	      dimension = beVertical ? "x":"y";
	      position = "0";
	      this.parent.putAxisIdPair(id, beVertical ? "primary-x":"primary-y");
	  }
	  else
	  {
		  dimension = beVertical ? "x":"y";
	      position = "end";
	      this.parent.putAxisIdPair(id, beVertical ? "secondary-x":"secondary-y");
	  }
      
      Number oRot = (Number)content.get("txRot");
      if(oRot!=null)
    	  this.rotationAngle = oRot.floatValue();
      JSONObject scaling = (JSONObject)content.get("scaling");
      if(scaling!=null)
      {
        orentation = (String)scaling.get("orientation");
        Number fLogBase = (Number)scaling.get("logBase");
        if(fLogBase!=null)
          logBase = fLogBase.floatValue();
        Number dMax = (Number)scaling.get("max");
        if(dMax != null)
        	max = dMax.doubleValue();
        Number dMin = (Number)scaling.get("min");
        if(dMin != null)
        	min = dMin.doubleValue();
      }
      
      Number majoru = (Number)content.get("majorUnit");
      if(majoru!=null)
        majorUnit = majoru.floatValue();
      
      Number minoru = (Number)content.get("minorUnit");
      if(minoru!=null)
        minorUnit = minoru.floatValue();
    
      JSONObject majorGridJson = (JSONObject)content.get("majorGridlines");
      if(majorGridJson!=null)
      {
        majorGrid = new Style.Graphic();
        majorGrid.loadFromJson(majorGridJson);
      }
      
      JSONObject minorGridJson = (JSONObject)content.get("minorGridlines");
      if(minorGridJson!=null)
      {
        minorGrid = new Style.Graphic();
        minorGrid.loadFromJson(minorGridJson);
      }
      
      JSONObject titleJson = (JSONObject)content.get("title");
      if(titleJson!=null)
      {
        title = new Title(chart);
        title.loadFromJson(titleJson);
      }
      
      JSONObject graphicJson = (JSONObject)content.get("spPr");
      if(graphicJson!=null)
      {
        graphicPro = new Style.Graphic();
        graphicPro.loadFromJson(graphicJson);
      }
      
      JSONObject textJson = (JSONObject)content.get("txPr");
      JSONObject chartTxPr = chart.getTextPro();
      if(textJson!=null)
      {
        textPro = new Style.Text();
        textPro.loadFromJson(textJson, chartTxPr);
      }
      
      JSONObject catJson = (JSONObject)content.get("cat");
      if(catJson!=null)
      {
        categoriesRef = new DataReference();
        categoriesRef.loadFromJson(catJson);
      }
      
      JSONObject numFmt = (JSONObject)content.get("numFmt");
      if(numFmt != null)
      {
    	  JSONObject formatJson = (JSONObject)numFmt.get("format");
    	  if(formatJson != null)
    		  format = new ChartNumberFormat(formatJson);
    	  
    	  Number sl = (Number)numFmt.get("sourceLinked");
    	  if(sl != null)
    		  sourceLinked = sl.intValue() == 1;
      }
      
      Number cha = (Number)content.get("changes");
	   if(cha != null)
		   changes = cha.intValue();
	}
	
	public String getOdfId()
	{
	  return this.parent.getOdfAxisId(id);
	}
}
